v7/navstories/navstories.py

```diff
@@ -26,24 +26,146 @@
+                nav_config[i] = utils.TranslatableSetting(i, {}, site.config['TRANSLATIONS'])
+
         site.scan_posts()
         # NAVIGATION_LINKS is a TranslatableSetting, values is an actual dict
         for lang in site.config['NAVIGATION_LINKS'].values:
-            new = []
+            # navstories config for lang
+            nav_conf_lang = {}
+            for i in self.conf_vars:
+                 nav_conf_lang[i] = nav_config[i].values[lang]
+
+            # Which paths are navstories active for current lang? - Must start and end with /
+            paths = tuple(('/' + s.strip('/') + '/') for s in nav_conf_lang['NAVSTORIES_PATHS'])
+
+            new_raw = {} # Unusorted (raw) new entries, deleted as mapped to new
+            new = [] # Sorted entries as a list of top-level menu entries, later
+            # Map site pages to new_raw structure
             for p in site.pages:
+                # Generate mavpath (menu) based on permalink without language prefix
+                # If TRANSLATION[DEFAULT_LANG] = '', then "permalink_nolang = p.permalink()" is ok
+                permalink_nolang = re.sub(r'^/' + nav_conf_lang['TRANSLATIONS'].lstrip('./') + '/?', '/', p.permalink(lang))
+                s_candidates = [s for s in paths if permalink_nolang.startswith(s)]
+                if not s_candidates:
+                    continue
+                # get longest path
+                s = max(s_candidates, key=len)
+                # Strip off the longest path in paths
+                navpath = permalink_nolang[len(s):].strip('/').split('/')
+                if len(navpath) == 0:
+                    # Should not happen that navpath is empty, but to prevent errors, and inform via a warning
+                    LOGGER.warn("Page with permalink: '%s', title: '%s', not added to menu by navstories.\033[0m" % (p.permalink(lang), p.title(lang)))
+                    continue
                 if lang in p.translated_to and not p.meta('hidefromnav'):
-                    new.append(p)
-            new_entries = tuple(sorted([(p.permalink(lang), p.title(lang)) for p in new]))
+                    # Add entry
+                    if not navpath[0] in new_raw:
+                        new_raw[navpath[0]] = []
+                    new_raw[navpath[0]].append([navpath, p.permalink(lang), p.title(lang)])
+
+            # Map from new_raw to new, sorting by NAVSTORIES_MAPPING
+            for map_key, map_txt in nav_conf_lang['NAVSTORIES_MAPPING']:
+                # Loop over all new_raw entries, checking if it matches map_key; if match: add it and delete from new_raw
+                if map_key in new_raw:
+                    new.append([map_txt, new_raw[map_key]])
+                    del(new_raw[map_key])
+            # Add remaing new_raw entries which didn't match any map_key
+            new.extend([[None, new_raw[_]] for _ in sorted(new_raw)])
 
+            # Map to tuble
+            new_entries = self.map_to_menu(new)
```

Comment:
I will change new to a OrderedDict where the key is the title and the value is an array of NavNode's.